# HW1: Các câu lệnh Linux, Python

(Cập nhật lần cuối 16/11/2021)

Họ tên: Nguyễn Ngọc Uyên Trang

MSSV: 19127074

---

## Cách làm bài và nộp bài

&#9889; Bạn lưu ý là mình sẽ dùng chương trình hỗ trợ chấm bài nên bạn cần phải tuân thủ chính xác qui định mà mình đặt ra, nếu không rõ thì hỏi, chứ không nên tự tiện làm theo ý của cá nhân.

**Cách làm bài**

Bạn sẽ làm trực tiếp trên file notebook này. Đầu tiên, bạn điền họ tên và MSSV vào phần đầu file ở bên trên. Trong file, bạn làm bài ở những chỗ có ghi là:
```python
# YOUR CODE HERE
raise NotImplementedError()
```
hoặc đối với những phần code không bắt buộc thì là:
```python
# YOUR CODE HERE (OPTION)
```
hoặc đối với markdown cell thì là:
```markdown
YOUR ANSWER HERE
```
Tất nhiên, khi làm thì bạn xóa dòng `raise NotImplementedError()` đi.
Đối những phần yêu cầu code thì thường ở ngay phía dưới sẽ có một (hoặc một số) cell chứa các bộ test để giúp bạn biết đã code đúng hay chưa; nếu chạy cell này không có lỗi gì thì có nghĩa là qua được các bộ test. Trong một số trường hợp, các bộ test có thể sẽ không đầy đủ; nghĩa là, nếu không qua được test thì là code sai, nhưng nếu qua được test thì chưa chắc đã đúng hoàn toàn.

Trong khi làm bài, bạn có thể cho in ra màn hình, tạo thêm các cell để test. Nhưng khi nộp bài thì bạn xóa các cell mà bạn tự tạo, xóa hoặc comment các câu lệnh in ra màn hình. Bạn lưu ý <font color=red>không được tự tiện xóa các cell hay sửa code của Thầy</font> (trừ những chỗ được phép sửa như đã nói ở trên).

Bạn cố gắng giữ cho code rõ ràng, ngăn nắp: đặt tên biến gợi ý, comment ở những chỗ cần comment, giữ cho một dòng code không quá dài (vì nếu không thì người đọc sẽ phải mất công scroll, bạn có thể dùng qui tắc: một dòng code chỉ nên có tối đa là 79 ký tự). 

Trong khi làm bài, thường xuyên `Ctrl + S` để lưu lại bài làm của bạn, tránh mất mát thông tin.


*Nên nhớ mục tiêu chính ở đây là <font color=green>học, học một cách chân thật</font>. Bạn có thể thảo luận ý tưởng với bạn khác, nhưng <font color=green>code và bài làm phải là của bạn, dựa trên sự hiểu thật sự của bạn</font>. <font color=red>Nếu vi phạm thì sẽ bị 0 điểm cho toàn bộ môn học.</font>*

**Cách nộp bài**

Khi chấm bài, đầu tiên mình sẽ chọn `Kernel` - `Restart & Run All` để restart và chạy tất cả các cell trong notebook của bạn; do đó, trước khi nộp bài, bạn nên chạy thử `Kernel` - `Restart & Run All` để đảm bảo mọi chuyện diễn ra đúng như mong đợi.

Sau đó, bạn tạo thư mục nộp bài theo cấu trúc sau:
- Thư mục `MSSV` (vd, nếu bạn có MSSV là 1234567 thì bạn đặt tên thư mục là `1234567`)
    - File `HW1.ipynb` (không cần nộp các file khác)

Cuối cùng, bạn nén thư mục `MSSV` này lại và nộp ở link trên moodle. <font color=red>Bạn lưu ý tuân thủ chính xác qui định nộp bài.</font>

---

## Môi trường code

Ta thống nhất trong môn này:
- Code trên môi trường Linux (bạn đã setup ở HW0)
- Dùng Jupyter Notebook hoặc Jupyter Lab để code file .ipynb. Lưu ý: bạn <font color=red>không dùng Google Colab</font>, vì Google Colab có thể xóa một số thông tin meta mà chương trình hỗ trợ chấm bài của mình đã tạo trong file bài tập
- Dùng phiên bản các package như trong file "min_ds-env.yml" (bạn xem trong HW0 cách dùng Conda để tạo môi trường code từ file .yml và cách kích hoạt môi trường này sau khi đã tạo)

In [1]:
import sys
sys.executable

'/home/uyen_trang/miniconda3/envs/min_ds-env/bin/python'

Nếu không có vấn đề gì thì file chạy python sẽ là file của môi trường code "min_ds-env".

Trong bài này, ngoài Git & Jupyter Notebook/Lab thì bạn chỉ dùng Python, không dùng các thư viện khác như Numpy, Pandas, ... Trong Python, khi đọc/ghi file csv thì bạn cũng không được dùng các thư viện hỗ trợ như thư viện `csv`.

---

## Import các thư viện cần thiết

In [2]:
import datetime
# YOUR CODE HERE (OPTION)


---

## Thu thập dữ liệu (3.5đ)

Trong bài tập này, bạn sẽ lấy dữ liệu log của [thùng chứa nbdime](https://github.com/KienTrann/nbdime). Nbdime là công cụ giúp Git có thể diff và merge file Notebook một cách phù hợp (thay vì chỉ coi như là một file text thông thường). Link mình cung cấp không phải là link đến thùng chứa gốc của nbdime mà là thùng chứa mình "fork" (đại khái là tạo ra một bản copy) từ thùng chứa gốc. Lý do: mình muốn cố định lại dữ liệu log để có thể cung cấp các bộ test để bạn phần nào biết là đã code đúng hay chưa (còn nếu dùng thùng chứa gốc thì dữ liệu log có thể sẽ thay đổi trong vài ngày tới, các bộ test lúc đó có thể sẽ không còn đúng nữa).

Để có thể lấy dữ liệu log thì việc đầu tiên bạn cần làm ở cell bên dưới là: 

- Dùng câu lệnh Git để clone thùng chứa về; bạn clone vào một thư mục cùng cấp với file notebook này. <font color=gray>Gợi ý: trong Jupyter Notebook, ở code cell, bạn có thể thực hiện một câu lệnh Linux (bao gồm câu lệnh Git) bằng cách thêm dấu ! ở đầu.</font>
- Sau đó, để xem thử thư mục vừa clone có nặng không, bạn tính dung lượng tổng cộng của thư mục này và lưu vào biến `repo_size` (`repo_size` là một con số, đơn vị là MB); ta thống nhất: (i) dùng cờ `--apparent-size` khi dùng câu lệnh `du`, (ii) code chỉ cần chạy đúng với thùng chứa cụ thể này. <font color=gray>Gợi ý: trong Jupyter Notebook, bạn có thể gán kết quả của câu lệnh Linux vào biến Python.</font> \
@ Các bạn dùng Mac: trong Mac thì mình nghe nói là câu lệnh `du` không có cờ `--apparent-size` (do câu lệnh `du` của Mac là bản BSD, còn của Linux là bản GNU); bạn có thể xem thử cách cài câu lệnh `du` bản GNU ở [video này](https://youtu.be/VGgTmxXp7xQ?t=894) (trong video là nói về câu lệnh `grep` nhưng mình nghĩ cũng sẽ tương tự với câu lệnh `du`).

In [3]:
# YOUR CODE HERE
#raise NotImplementedError()
!git clone https://github.com/KienTrann/nbdime.git
repo_size=!du -sh --apparent-size
repo_size
# Thầy HDTH với mình có test thử thì có máy ra 9.7, có máy ra 9.8
# (có thể là do phiên bản của câu lệnh du?)

Cloning into 'nbdime'...
remote: Enumerating objects: 12389, done.
remote: Counting objects: 100% (376/376), done.
remote: Compressing objects: 100% (256/256), done.
remote: Total 12389 (delta 198), reused 234 (delta 117), pack-reused 12013
Receiving objects: 100% (12389/12389), 5.21 MiB | 3.11 MiB/s, done.
Resolving deltas: 100% (9015/9015), done.


['11M\t.']

Sau khi đã clone thùng chứa về, việc kế tiếp của bạn là dùng các câu lệnh Linux (trong đó có các câu lệnh Git) và Python để lấy dữ liệu log rồi lưu xuống file "log.tsv" (định dạng tsv tương tự như csv nhưng các giá trị được cách nhau bởi tab). Để dễ hình dung, bạn có thể dùng Jupyter Lab để mở xem file kết quả đúng "correct_log.tsv" mà mình cung cấp. Như bạn có thể thấy, mỗi dòng ứng với một commit (không tính dòng đầu - dòng chứa các tên cột) và gồm các thông tin:
- "Id" là id (hash) dạng viết gọn của commit
- "ParentIds" là id dạng viết gọn của (các) commit cha
- "AuthorName" là tên của tác giả
- "AuthorEmail" là email của tác giả
- "AuthorDate" là ngày tác giả thực hiện commit, ngày được format theo chuẩn RFC2822 (xem phần gợi ý ở bên dưới để biết cách lấy ngày theo format này) 
- "Subject" là tựa đề của message (message tương tự như email, ngoài phần "subject" thì có thể có thêm phần "body"; ở đây ta chỉ lấy phần "subject")
- "ChangedFiles": các file bị thay đổi, nếu có nhiều file thì các file được phân tách bởi dấu phẩy

Tại sao mình dùng định dạng tsv thay vì csv? Vì dấu phẩy nhiều khả năng là sẽ có trong phần "subject" của message. Tab ít khả năng sẽ có hơn, nhưng nếu có thì bạn hãy thay bằng bốn khoảng trắng.

<font color=gray>Gợi ý cho phần này:</font> 
- <font color=gray>Bạn có thể xem thêm về câu lệnh `git log` [ở đây](https://git-scm.com/book/en/v2/Git-Basics-Viewing-the-Commit-History)</font>
- <font color=gray>Khi ở Notebook, bạn thử `cd` đến thư mục clone rồi `pwd` xem? Kết quả như vậy là vì khi thực hiện một dòng lệnh Linux ở Notebook thì Notebook sẽ tạo ra một terminal tạm thời để thực thi dòng lệnh đó trong phút chốc. Như vậy, câu lệnh `cd` vẫn là có `cd` đến thư mục clone nhưng là ở terminal tạm thời; khi câu lệnh `cd` chạy xong thì terminal tạm thời này sẽ được đóng và ta vẫn ở thư mục mà hiện chứa file Notebook. Để có thể `cd` rồi `pwd` thì một cách là ghi 2 câu lệnh này trên một dòng và cách nhau bởi `&&`; lúc này thì cả 2 câu lệnh sẽ được thực thi trên cùng một terminal</font>

In [4]:
# YOUR CODE HERE
#raise NotImplementedError()
def write_file(file_name,commit_list):
    with open(file_name, 'w') as file:
        for commit in commit_list:
            for word in commit[0:-1]:
                file.write(word + '\t')
            file.write(commit[-1])
                
            file.write('\n')
    
def get_commit_list():
    !cd nbdime 
    commit_list =[['Id','ParentIds','AuthorName','AuthorEmail','AuthorDate',
                   'Subject','ChangedFiles']]
    temp =!cd nbdime && git log --pretty=format:"%h~%p~%an~%ae~%aD~%s"
    for commit in temp:
        commit=commit.split('~')
        file_changed=[]
        if(('Merge'and'pull') not in commit[5] and ('Merge'and'branch')not in commit[5]): #Các subject là 'Merge pull' hoặc 'Merge branch' thì không có file thay đổi        
            check = !cd nbdime && git show --pretty="format:" {commit[0]}
            if(check):
                file_changed = !cd nbdime && git show --pretty="format:" --name-only {commit[0]}
            file_changed =','.join(file_changed) #các file được phân tách bởi dấu phẩy
        else: file_changed=''
        commit.append(file_changed)
        commit_list.append(commit)
    return commit_list

commit_list=get_commit_list()
write_file('log.tsv',commit_list)           


In [5]:
# TEST
file = open('log.tsv', 'r')
log = file.read()
file.close()
file = open('correct_log.tsv', 'r')
correct_log = file.read()
file.close()
assert log == correct_log

---

## Khám phá dữ liệu

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột? (0.5đ)

Đầu tiên, bạn sẽ đọc dữ liệu từ file "log.tsv" (nếu phần thu thập dữ liệu bạn làm chưa đúng thì bạn có thể đọc từ file "correct_log.tsv" mà mình đã cung cấp). Bạn lưu kết quả đọc được vào biến `cols`. Biến `cols` là một dictionary, trong đó mỗi phần tử ứng với một cột, key là chuỗi ứng với tên cột, value là list chứa các giá trị của cột (mỗi giá trị bạn cứ tạm để kiểu dữ liệu là chuỗi).

Kế tiếp, từ biến `cols`, bạn sẽ tính số dòng và số cột, rồi lần lượt lưu vào hai biến là `num_rows` và `num_cols`.

In [6]:
# YOUR CODE HERE
#raise NotImplementedError()
def get_cols():
    file = open('log.tsv', 'r')
    Lines = file.readlines()
    cols={}
    keys = Lines[0].strip().split('\t')
    for i in keys:
        cols[i] =[]
    for line in Lines[1:]:
        line = line.strip().split('\t')
        cols['Id'].append(line[0])
        cols['ParentIds'].append(line[1])
        cols['AuthorName'].append(line[2])
        cols['AuthorEmail'].append(line[3])
        cols['AuthorDate'].append(line[4])
        cols['Subject'].append(line[5])
        try:
            cols['ChangedFiles'].append(line[6])
        except: 
            cols['ChangedFiles'].append('')
    return cols
cols = get_cols()
num_rows = len(cols['Id'])
num_cols=len(cols)
num_rows,num_cols

(1928, 7)

In [7]:
# TEST
assert num_rows == 1928
assert num_cols == 7
assert set(cols.keys()) == set(['Id', 'ParentIds', 'AuthorName', 
                                'AuthorEmail', 'AuthorDate', 'Subject', 
                                'ChangedFiles'])
id_col_len = len(cols['Id'])
for col_name in cols:
    assert len(cols[col_name]) == id_col_len

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng ứng với một commit. Có vẻ không có dòng nào "lạc loài" (nghĩa là không có vấn đề các dòng có ý nghĩa khác nhau).

### Dữ liệu có các dòng bị lặp không? (0.25đ)

Có lẽ là không có. Để chắc ăn thì một cách là ta có thể check cột "Id"; nếu cột "Id" không có giá trị bị lặp thì có thể kết luận là không có dòng bị lặp. Bạn sẽ làm vụ này và lưu kết quả check vào biến `duplicated_id`; biến này bằng True nếu cột "Id" có giá trị lặp và bằng False nếu ngược lại.

In [8]:
# YOUR CODE HERE
#raise NotImplementedError()
Id_set = set(cols['Id'])
if(len(Id_set) == len(cols['Id'])):
    duplicated_id= False
else: duplicated_id=True

In [9]:
# TEST
assert duplicated_id == False

### Mỗi cột có ý nghĩa gì?

Vụ này mình đã nói lúc thu thập dữ liệu. Ở đây, mình ghi lại một lần nữa:

- "Id" là id (hash) dạng viết gọn của commit
- "ParentIds" là id dạng viết gọn của (các) commit cha
- "AuthorName" là tên của tác giả
- "AuthorEmail" là email của tác giả
- "AuthorDate" là ngày tác giả thực hiện commit
- "Subject" là tựa đề của message (message tương tự như email, ngoài phần "subject" thì có thể có thêm phần "body"; ở đây ta chỉ lấy phần "subject")
- "ChangedFiles": các file bị thay đổi, nếu có nhiều file thì các file được phân tách bởi dấu phẩy

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

- Các cột "Id", "ParentIds", "AuthorName", "AuthorEmail", "Subject": kiểu dữ liệu dạng chuỗi là đúng rồi
- Cột "ChangedFiles": hmm ... có lẽ tạm thời ta vẫn để dạng chuỗi, có gì cần thì ta sẽ xử lý sau
- Cột "AuthorDate": cần chuyển từ kiểu dữ liệu dạng chuỗi sang datetime; để có thể tiếp tục khám phá tiếp thì ngay sau đây bạn sẽ tiền xử lý vụ này luôn

---

## Tiền xử lý (0.5đ)

Ở đây, bạn sẽ chuyển cột "AuthorDate" từ dạng chuỗi sang datetime. <font color=gray>Gợi ý: bạn xem [document](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes) để biết cách truyền format phù hợp cho hàm `strptime`.</font>

In [10]:
# YOUR CODE HERE
#raise NotImplementedError()
for i in range (len(cols['AuthorDate'])):
    cols['AuthorDate'][i]=datetime.datetime.strptime(cols['AuthorDate'][i], '%a, %d %b %Y %H:%M:%S %z')

In [11]:
# TEST
assert type(cols['AuthorDate'][0]) == datetime.datetime
assert cols['AuthorDate'][0] == datetime.datetime(2021, 4, 20, 12, 27, 9, 
            tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))

---

## Khám phá dữ liệu (tiếp tục)

### Với mỗi cột có kiểu dữ liệu dạng số (numeric) và dạng datetime, các giá trị được phân bố như thế nào? (0.5đ)

Ở đây thì không có cột dạng số, chỉ có một cột dạng datetime là cột "AuthorDate".

Đầu tiên, bạn tính tỉ lệ % các giá trị thiếu của cột "AuthorDate". <font color=gray>Gợi ý: không cần code, có thể lập luận dựa vào kết quả khi chuyển cột "AuthorDate" từ dạng chuỗi sang datetime ở bên trên.</font>

YOUR ANSWER HERE

Do kết quả khi chuyển cột "AuthorDate" từ dạng chuỗi sang `datetime` ở bên trên không xảy ra lỗi, tức không có giá trị `' '` nên tỉ lệ % các giá trị thiếu của cột "AuthorDate" = 0%


Tiếp theo, bạn tính min và max của cột "AuthorDate", rồi lần lượt lưu vào hai biến `author_date_min` và `author_date_max`.

In [12]:
# YOUR CODE HERE
#raise NotImplementedError()

author_date_min=min(cols['AuthorDate'])
author_date_max=max(cols['AuthorDate'])

In [13]:
# TEST
assert author_date_min == datetime.datetime(2015, 11, 9, 13, 13, 28, 
            tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))
assert author_date_max == datetime.datetime(2021, 4, 20, 12, 27, 9, 
            tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))

Hmm ... Min và max của cột "AuthorDate" không có gì bất thường.

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào? (1.75đ)

Các cột dạng phân loại là các cột mà không phải là cột "AuthorDate". Bạn sẽ lưu thông tin của các cột này vào biến `cate_col_profiles`. Biến này là một dictionary, trong đó mỗi phần tử ứng với một cột dạng phân loại, key là tên cột, value là list gồm: tỉ lệ % giá trị thiếu, số lượng các giá trị khác nhau (không tính giá trị thiếu), set chứa các giá trị khác nhau (không bao gồm giá trị thiếu).

In [14]:
# YOUR CODE HERE
#raise NotImplementedError()
def count_missing(cols_key):
    missing = 0
    for value in cols_key:
        if(value == ''):
            missing +=1
    return missing
def get_cate_col_profiles(cols):
    keys_subset=['Id', 'ParentIds', 'AuthorName', 'AuthorEmail','Subject', 'ChangedFiles']
    cate_col_profiles=dict.fromkeys(keys_subset,None)
    for key in cate_col_profiles.keys():
        cate_col_profiles[key]=[]
        missing = count_missing(cols[key])
        percent_missing = (missing / len(cols[key]))*100
        temp_set = set(cols[key])
        temp_set.discard('')
        nums_set = len(temp_set)
        cate_col_profiles[key].append(percent_missing)
        cate_col_profiles[key].append(nums_set)
        cate_col_profiles[key].append(temp_set)
    return cate_col_profiles
cate_col_profiles = get_cate_col_profiles(cols)

In [15]:
# TEST
assert set(cate_col_profiles.keys()) == {'AuthorEmail', 'AuthorName', 
                                         'ChangedFiles', 'Id', 
                                         'ParentIds', 'Subject'}
for col_name, col_profile in cate_col_profiles.items():
    assert len(col_profile) == 3
    assert len(col_profile[2]) == col_profile[1]
assert round(cate_col_profiles['ParentIds'][0], 3) == 0.052
assert round(cate_col_profiles['ChangedFiles'][0], 3) == 16.079
assert cate_col_profiles['ParentIds'][1] == 1820
assert cate_col_profiles['ChangedFiles'][1] == 842

In [16]:
# TEST (BY EYES)
print(f"{'ColName':12} {'Miss(%)':7} {'NumDifVals':10} {'SomeVals'}")
for col_name, col_profile in cate_col_profiles.items():
    print(f'{col_name:12} {col_profile[0]:<7.3f} {col_profile[1]:<10} {col_profile[2].__repr__()[:34]+"...":<}')

ColName      Miss(%) NumDifVals SomeVals
Id           0.000   1928       {'4659b81', 'f014b0f', '26a67d4', ...
ParentIds    0.052   1820       {'4659b81', 'f014b0f', '26a67d4', ...
AuthorName   0.000   34         {'Charlotte Godley', 'Jonathan Whi...
AuthorEmail  0.000   34         {'nbollweg@continuum.io', 'fcollon...
Subject      0.000   1860       {'Add failing test for config inhe...
ChangedFiles 16.079  842        {'packages/nbdime/src/diff/util.ts...


Bạn quan sát kết quả in ra ở trên (tạm thời chưa để ý đến cột "Miss(%)") và cho biết có gì bất thường không? Hay mọi thứ đều bình thường? Cho mình biết tại sao bạn lại kết luận như vậy.

YOUR ANSWER HERE

Cột `Subject` không có ô trống nhưng số lượng subject khác nhau là `1860` < `1928` (tổng số lượng subject) -> có `1928-1860=68` subject giống nhau

Kế tiếp, bạn hãy cho in ra các dòng mà có giá trị thiếu (in sao là tùy bạn, miễn là dễ nhìn). Sau đó, bạn hãy quan sát các dòng này để xem tại sao lại bị thiếu (trong trường hợp, bạn không biết tại sao thì cứ ghi nhận là không biết tại sao).

In [17]:
# YOUR CODE HERE
#raise NotImplementedError()
print(f"{'Id':7}{'ParentIds':9} {'AuthorName':6} {'AuthorEmail':7} {'AuthorDate':12} {'Subject':13}{'ChangedFiles'}")

for commit in commit_list:
    for value in commit:
        if(value ==''):
            print(f'{commit[0]:8}{commit[1][:7]+"...":10}{commit[2][:6]+"...":11}{commit[3][:7]+"...":12}{commit[4][:11]+"...":11} {commit[5][:11]+"...":11} {commit[6][:10]+"..."}')
            break

Id     ParentIds AuthorName AuthorEmail AuthorDate   Subject      ChangedFiles
524aac2 2994fb4...Vidar ...  vidartf...  Tue, 20 Apr... Merge pull ... ...
ab8e60d f1e18ff...Vidar ...  vidartf...  Wed, 14 Apr... Merge pull ... ...
f1e18ff 52f7990...Vidar ...  vidartf...  Wed, 14 Apr... Merge pull ... ...
52f7990 8ce708c...Vidar ...  vidartf...  Tue, 13 Apr... Merge pull ... ...
8ce708c 4291b8c...Vidar ...  vidartf...  Mon, 12 Apr... Merge pull ... ...
075f502 2d1edb8...Vidar ...  vidartf...  Sun, 11 Apr... Merge branc... ...
4291b8c ab9f1bc...Vidar ...  vidartf...  Sun, 11 Apr... Merge pull ... ...
ab9f1bc c823623...Vidar ...  vidartf...  Sun, 11 Apr... Merge pull ... ...
fe2911a c38cfc5...Michał...  5832902...  Sun, 7 Mar ... Merge pull ... ...
a0726e5 75363a0...Vidar ...  vidartf...  Thu, 14 Jan... Merge pull ... ...
75363a0 f6beb60...Vidar ...  vidartf...  Thu, 7 Jan ... Merge pull ... ...
f6beb60 92fc22d...Vidar ...  vidartf...  Fri, 2 Oct ... Merge pull ... ...
77cac58 b707946...Vid

YOUR ANSWER HERE

-Các commit là commit cha nên sẽ không có `ParentIds`

-Các commit có `Subject` là `Merge pull` hoặc `Merge branch` thì không có file thay đổi

---

## Đưa ra câu hỏi có ý nghĩa cần trả lời

Sau khi khám phá dữ liệu, ta đã hiểu hơn về dữ liệu. Bây giờ, ta hãy xem thử có câu hỏi nào có ý nghĩa mà có thể được trả lời bằng dữ liệu này không.

Các câu hỏi có thể có là: 
- *Câu hỏi 1: trong năm 2021, ai là người có nhiều commit nhất, ai là người có nhiều commit nhì, ...?* \
Tìm câu trả lời cho câu hỏi này để làm gì? \
Hmm ... Chẳng hạn để khen thưởng
- *Câu hỏi 2: commit nào là commit cuối cùng (xét trên "AuthorDate") đã thay đổi file "nbdime/webapp/templates/difftool.html"?* \
Tìm câu trả lời cho câu hỏi này để làm gì? \
Hmm ... Chẳng hạn để đổ tội cho tác giả của commit đó ;-)

---

## Tiền xử lý (nếu cần) + phân tích dữ liệu để trả lời cho câu hỏi 1 (1đ)

Nhắc lại câu hỏi 1: *trong năm 2021, ai là người có nhiều commit nhất, ai là người có nhiều commit nhì, ...?*

Bạn sẽ lưu kết quả vào biến `wanted_people`. Biến này là một dictionary, mỗi phần tử ứng với một người mà có commit trong năm 2021, key là tên của người đó, value là số lượng commit của người đó. Dictionary này được sort theo value giảm dần.

In [18]:
# YOUR CODE HERE
#raise NotImplementedError()
wanted_people={}
for i in range (len(cols['Id'])):
    if(cols['AuthorDate'][i].year==2021):
        wanted_people[cols['AuthorName'][i]]=wanted_people.get(cols['AuthorName'][i],0)+1
wanted_people=dict(sorted(wanted_people.items(), key = lambda x: x[1], reverse = True))
wanted_people

{'Vidar Tonaas Fauske': 26,
 'Alex Bozarth': 9,
 'Frederic COLLONVAL': 7,
 'Frédéric Collonval': 6,
 'krassowski': 2,
 'Michał Krassowski': 2}

In [19]:
# TEST
correct_wanted_people = {'Vidar Tonaas Fauske': 26, 
                         'Alex Bozarth': 9, 
                         'Frederic COLLONVAL': 7, 
                         'Frédéric Collonval': 6, 
                         'krassowski': 2, 
                         'Michał Krassowski': 2}
assert wanted_people == correct_wanted_people

---

## Tiền xử lý (nếu cần) + phân tích dữ liệu để trả lời cho câu hỏi 2 (2đ)

Nhắc lại câu hỏi 2: *commit nào là commit cuối cùng (xét trên "AuthorDate") đã thay đổi file "nbdime/webapp/templates/difftool.html"?*

Ta thống nhất: không xét các commit "merge pull request" vì các commit này không có thông tin về "ChangedFiles".

Bạn lưu kết quả vào biến `wanted_commit`. Biến này là một dictionary, trong đó mỗi phần tử ứng với một thông tin của commit, key là tên cột, value là giá trị tương ứng.

In [20]:
# YOUR CODE HERE
#raise NotImplementedError()
wanted_commit={}
max_date=-1
for i in range (len(cols['Id'])):
    if('nbdime/webapp/templates/difftool.html' in cols['ChangedFiles'][i]):
        if(cols['AuthorDate'][i]>cols['AuthorDate'][max_date]):
            max_date=i
for key,value in cols.items():
    wanted_commit[key] = value[max_date]
wanted_commit

{'Id': 'be0b5ee',
 'ParentIds': '4226179',
 'AuthorName': 'Vidar Tonaas Fauske',
 'AuthorEmail': 'vidartf@gmail.com',
 'AuthorDate': datetime.datetime(2017, 7, 4, 22, 53, 9, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
 'Subject': 'jinja2 templates + avoid name collisions with nb',
 'ChangedFiles': 'nbdime/webapp/nbdimeserver.py,nbdime/webapp/templates/compare.html,nbdime/webapp/templates/diff.html,nbdime/webapp/templates/difftool.html,nbdime/webapp/templates/merge.html,nbdime/webapp/templates/mergetool.html,nbdime/webapp/templates/nbdimepage.html,setup.py'}

In [21]:
# TEST
correct_wanted_commit = {\
    'Id': 'be0b5ee',
    'ParentIds': '4226179',
    'AuthorName': 'Vidar Tonaas Fauske',
    'AuthorEmail': 'vidartf@gmail.com',
    'AuthorDate': datetime.datetime(2017, 7, 4, 22, 53, 9, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'Subject': 'jinja2 templates + avoid name collisions with nb',
    'ChangedFiles': 'nbdime/webapp/nbdimeserver.py,nbdime/webapp/templates/compare.html,nbdime/webapp/templates/diff.html,nbdime/webapp/templates/difftool.html,nbdime/webapp/templates/merge.html,nbdime/webapp/templates/mergetool.html,nbdime/webapp/templates/nbdimepage.html,setup.py'\
}
assert wanted_commit == correct_wanted_commit